# Загрузка Pandas и очистка данных

In [197]:
import pandas as pd
import numpy as np

In [198]:
df = pd.read_csv('main_task.csv')

In [199]:
df.head()

Restaurant_id       City                                      Cuisine Style  \
0       id_5569      Paris            ['European', 'French', 'International']   
1       id_1535  Stockholm                                                NaN   
2        id_352     London  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...   
3       id_3456     Berlin                                                NaN   
4        id_615     Munich  ['German', 'Central European', 'Vegetarian Fri...   

   Ranking  Rating Price Range  Number of Reviews  \
0   5570.0     3.5    $$ - $$$              194.0   
1   1537.0     4.0         NaN               10.0   
2    353.0     4.5        $$$$              688.0   
3   3458.0     5.0         NaN                3.0   
4    621.0     4.0    $$ - $$$               84.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   

                                              URL_TA     ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...  d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...  d6864963

In [200]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  object 
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  37457 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), object(7)
memory usage: 3.1+ MB


Приведем имена колонок к нормальному виду

In [201]:
df.columns=['restaurant_id', 'city', 'cuisine_style', 'ranking', 'rating',
       'price_range', 'reviews_cnt', 'reviews', 'ta_url', 'ta_id']


### Общие наблюдения и очистки

In [202]:
df.describe(include='all')

restaurant_id    city cuisine_style       ranking        rating  \
count          40000   40000         30717  40000.000000  40000.000000   
unique         11909      31          9007           NaN           NaN   
top           id_227  London   ['Italian']           NaN           NaN   
freq              18    5757          1032           NaN           NaN   
mean             NaN     NaN           NaN   3676.028525      3.993037   
std              NaN     NaN           NaN   3708.749567      0.668417   
min              NaN     NaN           NaN      1.000000      1.000000   
25%              NaN     NaN           NaN    973.000000      3.500000   
50%              NaN     NaN           NaN   2285.000000      4.000000   
75%              NaN     NaN           NaN   5260.000000      4.500000   
max              NaN     NaN           NaN  16444.000000      5.000000   

       price_range   reviews_cnt   reviews  \
count        26114  37457.000000     40000   
unique           3           NaN     33516   
top       $$ - $$$           NaN  [[], []]   
freq         18412           NaN      6471   
mean           NaN    124.825480       NaN   
std            NaN    295.666352       NaN   
min            NaN      2.000000       NaN   
25%            NaN      9.000000       NaN   
50%            NaN     33.000000       NaN   
75%            NaN    115.000000       NaN   
max            NaN   9660.000000       NaN   

                                                   ta_url     ta_id  
count                                               40000     40000  
unique                                              39980     39980  
top     /Restaurant_Review-g187514-d7809594-Reviews-La...  d7342803  
freq                                                    2         2  
mean                                                  NaN       NaN  
std                                                   NaN       NaN  
min                                                   NaN       NaN  
25%                                                   NaN       NaN  
50%                                                   NaN       NaN  
75%                                                   NaN       NaN  
max                                                   NaN       NaN

Есть неуникальные _restaurant_id_ и _ta_id_

In [203]:
df[df.restaurant_id==df.restaurant_id.value_counts().index[0]].head(3)



restaurant_id        city  \
1065        id_227  Luxembourg   
4727        id_227  Bratislava   
6237        id_227      Athens   

                                          cuisine_style  ranking  rating  \
1065  ['Lebanese', 'Mediterranean', 'Middle Eastern'...    228.0     4.0   
4727  ['Healthy', 'European', 'Mediterranean', 'Vege...    228.0     4.0   
6237             ['European', 'Greek', 'Mediterranean']    228.0     4.5   

     price_range  reviews_cnt  \
1065    $$ - $$$         21.0   
4727    $$ - $$$        122.0   
6237    $$ - $$$        220.0   

                                                reviews  \
1065  [['Nice little Lebanese restaurant', 'Great fo...   
4727  [['Great fish (African sharptooth catfish)', '...   
6237  [['Spectacular view and food', 'Great view'], ...   

                                                 ta_url      ta_id  
1065  /Restaurant_Review-g190356-d10275783-Reviews-K...  d10275783  
4727  /Restaurant_Review-g274924-d4009959-Reviews-Me...   d4009959  
6237  /Restaurant_Review-g189400-d1898490-Reviews-Ol...   d1898490

Одинаковый _restaurant_id_ встречается в разных городах и имеет разный _ta_id_. Удалим эту колонку.

In [204]:
df.drop(["restaurant_id"], axis=1, inplace=True)

In [205]:
with pd.option_context('display.max_colwidth', -1):
    print(df[df.ta_id==df.ta_id.value_counts().index[0]].head(3))

city                 cuisine_style  ranking  rating price_range  \
22924  Madrid  ['Spanish', 'Mediterranean']  2361.0   4.5     $$ - $$$     
36275  Madrid  ['Spanish', 'Mediterranean']  2469.0   4.5     $$ - $$$     

       reviews_cnt                                          reviews  \
22924  16.0         [['The real codido madrileño'], ['03/31/2017']]   
36275  16.0         [['The real codido madrileño'], ['03/31/2017']]   

                                                                    ta_url  \
22924  /Restaurant_Review-g187514-d7342803-Reviews-Los_Hierros-Madrid.html   
36275  /Restaurant_Review-g187514-d7342803-Reviews-Los_Hierros-Madrid.html   

          ta_id  
22924  d7342803  
36275  d7342803  


Судя по-всему дублирующиеся _ta_id_ отличаются только по _ranking_. Попробуем сгрупировать, при этом ранк возьмется случайно, что не страшно, т.к. дублей всего 20

In [206]:
df.drop_duplicates(subset=['city', 'cuisine_style', 'rating','price_range', 'reviews_cnt', 'reviews', 'ta_url', 'ta_id'], inplace=True, ignore_index=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39980 entries, 0 to 39979
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   city           39980 non-null  object 
 1   cuisine_style  30701 non-null  object 
 2   ranking        39980 non-null  float64
 3   rating         39980 non-null  float64
 4   price_range    26101 non-null  object 
 5   reviews_cnt    37437 non-null  float64
 6   reviews        39980 non-null  object 
 7   ta_url         39980 non-null  object 
 8   ta_id          39980 non-null  object 
dtypes: float64(3), object(6)
memory usage: 2.7+ MB


Теперь количество строк совпадает с уникальными значениями _ta_id_ т.е. его можно использовать как первичный ключ для дальнейших очисток

### Исследование  price_range

In [207]:
df.price_range.value_counts()

$$ - $$$    18402
$            6276
$$$$         1423
Name: price_range, dtype: int64

Это ординальный признак, но есть пропуски. Преобразуем в числа.

In [208]:
dct = {'$$ - $$$': 2, '$': 1, '$$$$': 3}
df.price_range.replace(to_replace=dct, inplace=True)

Пропуски заполним наиболее частым значением (2), т.к. оно более вероятно

In [209]:
df.price_range.fillna(2, inplace=True)

### Исследование и очистка city

In [210]:
df.city.describe()

count      39980
unique        31
top       London
freq        5757
Name: city, dtype: object

### Исследование и очистка cuisine

Сделаем вспомогательный датафрейм

In [211]:
cuisines = df[["ta_id","cuisine_style"]]

In [212]:
import json
cuisines.cuisine_style.fillna('["Unknown"]', inplace=True)
cuisines.cuisine_style = cuisines.cuisine_style.apply(lambda c: json.loads(c.replace('\'','\"')))

Преобразуем массив кухонь в отдельный ряд на кухню для каждого ресторана

In [213]:
cuisines = cuisines.explode("cuisine_style")
cuisines.describe()

ta_id        cuisine_style
count     104843               104843
unique     39980                  126
top     d7231902  Vegetarian Friendly
freq          21                11183

Число уникальных кухонь - 126, но без учета добавленной Unknown - 125

Самые популярные кухни:

In [214]:
cuisines.cuisine_style.value_counts()


Vegetarian Friendly    11183
European               10056
Unknown                 9279
Mediterranean           6271
Italian                 5963
                       ...  
Xinjiang                   1
Yunnan                     1
Salvadoran                 1
Latvian                    1
Burmese                    1
Name: cuisine_style, Length: 126, dtype: int64

Среднее число кухонь на ресторан:

In [215]:
cuisines.groupby("ta_id").nunique().mean()

ta_id            1.000000
cuisine_style    2.622386
dtype: float64

Добавим фичу с колличеством кухонь на ресторан в основной датасет

In [216]:
cuisine_cnt_df = cuisines.groupby("ta_id").nunique()
cuisine_cnt_df.columns=["ta_id","cuisines_cnt"]
cuisine_cnt_df = cuisine_cnt_df[["cuisines_cnt"]].reset_index()
cuisine_cnt_df
df=df.merge(cuisine_cnt_df, how="left", on="ta_id")
df.head(3)

city                                      cuisine_style  ranking  \
0      Paris            ['European', 'French', 'International']   5570.0   
1  Stockholm                                                NaN   1537.0   
2     London  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...    353.0   

   rating  price_range  reviews_cnt  \
0     3.5          2.0        194.0   
1     4.0          2.0         10.0   
2     4.5          3.0        688.0   

                                             reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   

                                              ta_url     ta_id  cuisines_cnt  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643             3  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032             1  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781             7

Развернем данные по кухням, взяв только топ 25\* кухонь. 

\*Проверено опытным путем что большее количество уже незначительно влияет на ошибку модели

In [217]:
top_kitchens = cuisines.cuisine_style.value_counts()[:25].index 
cuisines.cuisine_style=cuisines["cuisine_style"].apply(lambda c: c if c in top_kitchens else 'other')

In [218]:
cuisines=pd.get_dummies(cuisines,columns=["cuisine_style"]).groupby("ta_id").sum()

Смерджим с основным датасетом

In [219]:
df=df.merge(cuisines, how="left", on="ta_id")
df.drop("cuisine_style", axis=1, inplace=True)
df.head(3)

city  ranking  rating  price_range  reviews_cnt  \
0      Paris   5570.0     3.5          2.0        194.0   
1  Stockholm   1537.0     4.0          2.0         10.0   
2     London    353.0     4.5          3.0        688.0   

                                             reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   

                                              ta_url     ta_id  cuisines_cnt  \
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643             3   
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032             1   
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781             7   

   cuisine_style_American  ...  cuisine_style_Pizza  cuisine_style_Portuguese  \
0                       0  ...                    0                         0   
1                       0  ...                    0                         0   
2                       0  ...                    0                         0   

   cuisine_style_Pub  cuisine_style_Seafood  cuisine_style_Spanish  \
0                  0                      0                      0   
1                  0                      0                      0   
2                  0                      0                      0   

   cuisine_style_Sushi  cuisine_style_Unknown  cuisine_style_Vegan Options  \
0                    0                      0                            0   
1                    0                      1                            0   
2                    1                      0                            1   

   cuisine_style_Vegetarian Friendly  cuisine_style_other  
0                                  0                    0  
1                                  0                    0  
2                                  1                    1  

[3 rows x 36 columns]

In [220]:
df = df.drop([c for c in df.columns if df[c].dtype == 'object'], axis = 1) # Drop non-numbers



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Ranking            40000 non-null  float64
 1   Rating             40000 non-null  float64
 2   Number of Reviews  40000 non-null  float64
dtypes: float64(3)
memory usage: 937.6 KB


In [221]:
df["reviews_cnt"].fillna(df["reviews_cnt"].median(), inplace=True)

# Разбиваем датафрейм на части, необходимые для обучения и тестирования модели

In [222]:
# Х - данные с информацией о ресторанах, у - целевая переменная (рейтинги ресторанов)
X = df.drop([ 'rating'], axis = 1)
y = df['rating']

In [223]:
# Загружаем специальный инструмент для разбивки:
from sklearn.model_selection import train_test_split

In [224]:
# Наборы данных с меткой "train" будут использоваться для обучения модели, "test" - для тестирования.
# Для тестирования мы будем использовать 25% от исходного датасета.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

# Создаём, обучаем и тестируем модель

In [225]:
# Импортируем необходимые библиотеки:
from sklearn.ensemble import RandomForestRegressor # инструмент для создания и обучения модели
from sklearn import metrics # инструменты для оценки точности модели

In [226]:
# Создаём модель
regr = RandomForestRegressor(n_estimators=100)

# Обучаем модель на тестовом наборе данных
regr.fit(X_train, y_train)

# Используем обученную модель для предсказания рейтинга ресторанов в тестовой выборке.
# Предсказанные значения записываем в переменную y_pred
y_pred = regr.predict(X_test)

In [227]:
# Сравниваем предсказанные значения (y_pred) с реальными (y_test), и смотрим насколько они в среднем отличаются
# Метрика называется Mean Absolute Error (MAE) и показывает среднее отклонение предсказанных значений от фактических.
print('MAE:', metrics.mean_absolute_error(y_test, y_pred))

MAE: 0.3750028686962529
